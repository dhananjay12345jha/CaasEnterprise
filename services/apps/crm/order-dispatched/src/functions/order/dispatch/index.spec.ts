import { faker } from "@faker-js/faker";
import {
  default as triggerOrderDispatchedNotificationHandler,
  HandlerEvent,
} from "./index";
import { Context as LambdaContext } from "aws-lambda/handler";

const triggerNotificationMock = jest.fn();

jest.mock("./trigger-order-dispatched-notification", () => {
  const { OrderDispatched } = jest.requireActual(
    "./trigger-order-dispatched-notification"
  );
  return {
    OrderDispatched: jest.fn(() => {
      return {
        ...OrderDispatched,
        triggerNotification: triggerNotificationMock,
      };
    }),
  };
});

describe("trigger order dispatched notification handler", () => {
  const expectedResponse = {
    dispatch_id: faker.datatype.uuid(),
  };

  beforeEach(() => jest.restoreAllMocks());

  it("should throw an error if brand id is missing", async () => {
    const event: HandlerEvent = {
      version: "1.0",
      id: faker.datatype.string(),
      "detail-type": "order.shipment.changed",
      time: faker.datatype.string(),
      detail: {
        referenced: false,
        metadata: { "x-emc-ubid": null },
        correlationId: faker.datatype.uuid(),
        payload: {
          orderId: faker.datatype.string(),
          orderVersion: 4,
          previousShipmentState: "Ready",
          newShipmentState: "Shipped",
        },
      },
    };

    await expect(() =>
      triggerOrderDispatchedNotificationHandler(event, {} as LambdaContext)
    ).rejects.toMatchSnapshot();
  });

  it("should return a dispatched response body and passes all input fields", async () => {
    triggerNotificationMock.mockResolvedValue(expectedResponse);

    const brandId = faker.datatype.uuid();

    const event: HandlerEvent = {
      version: "1.0",
      id: faker.datatype.uuid(),
      "detail-type": "order.shipment.changed",
      time: faker.datatype.string(),
      detail: {
        referenced: false,
        metadata: { "x-emc-ubid": brandId },
        correlationId: faker.datatype.uuid(),
        payload: {
          orderId: faker.datatype.string(),
          orderVersion: 4,
          previousShipmentState: "Ready",
          newShipmentState: "Shipped",
        },
      },
    };

    const response = await triggerOrderDispatchedNotificationHandler(
      event,
      {} as LambdaContext
    );

    expect(response).toBe(expectedResponse);
    expect(triggerNotificationMock).toBeCalledTimes(1);
    expect(triggerNotificationMock).toBeCalledWith(
      brandId,
      event.detail.payload
    );
  });

  it("should throw an error if something went wrong", async () => {
    triggerNotificationMock.mockImplementation(() => {
      throw new Error();
    });

    const brandId = faker.datatype.uuid();

    const event: HandlerEvent = {
      version: "1.0",
      id: faker.datatype.uuid(),
      "detail-type": "order.shipment.changed",
      time: faker.datatype.string(),
      detail: {
        referenced: false,
        metadata: { "x-emc-ubid": brandId },
        correlationId: faker.datatype.uuid(),
        payload: {
          orderId: faker.datatype.string(),
          orderVersion: 4,
          previousShipmentState: "Ready",
          newShipmentState: "Shipped",
        },
      },
    };

    await expect(() =>
      triggerOrderDispatchedNotificationHandler(event, {} as LambdaContext)
    ).rejects.toThrow();
  });
});
