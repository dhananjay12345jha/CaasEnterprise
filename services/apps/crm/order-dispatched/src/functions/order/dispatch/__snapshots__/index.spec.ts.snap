// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trigger order dispatched notification handler should throw an error if brand id is missing 1`] = `
Array [
  Object {
    "code": "invalid_type",
    "expected": "string",
    "message": "Expected string, received null",
    "path": Array [
      "detail",
      "metadata",
      "x-emc-ubid",
    ],
    "received": "null",
  },
]
`;
